repo2docker/app.py

```diff
@@ -91,23 +91,62 @@ def _default_log_level(self):
     )
 
     def fetch(self, url, ref, checkout_path):
-        try:
-            for line in execute_cmd(['git', 'clone', url, checkout_path],
-                                    capture=self.json_logs):
-                self.log.info(line, extra=dict(phase='fetching'))
-        except subprocess.CalledProcessError:
-            self.log.error('Failed to clone repository!', extra=dict(phase='failed'))
-            sys.exit(1)
+        def _clone(depth=None):
+            if depth is not None:
+                command = ['git', 'clone', '--depth', str(depth),
+                           url, checkout_path]
+            else:
+                command = ['git', 'clone', url, checkout_path]
 
-        if ref:
             try:
-                for line in execute_cmd(['git', 'reset', '--hard', ref], cwd=checkout_path,
+                for line in execute_cmd(command, capture=self.json_logs):
+                    self.log.info(line, extra=dict(phase='fetching'))
+            except subprocess.CalledProcessError:
+                self.log.error('Failed to clone repository!',
+                               extra=dict(phase='failed'))
+                sys.exit(1)
```

Comment:
This should just re-raise, IMO. Calling sys.exit makes it hard for this to be used as a library, for example.

Reply:
On second thought, I wonder if this code can be simplified by not having them be functions. 

I generally feel that functions that are being called only from one place and without arguments shouldn't be functions - we can always abstract them when we realize we need to call them from two places. This means less mental jumping when reading the code, which is always good. 

In this context, there is also not much use for catching the CalledProcessError manually each time. execute_cmd will already output stderr, so there should already be a more specific error message anyway. So the generic 'failed to clone repository' can be a generic error handler for the entire fetch function, rather than for each mini-function.

So to summarize:

1. Remove the _ functions, just inline the code into fetch
2. Remove the individual try / excepts
3. ???
4. PROFIT

Reply:
There will be something on the screen, but I think it makes sense to have a "summary" error message. My guess would be that most users of `repo2docker` don't realise that the below message means that `fetch()` failed:
```
Cloning into '/tmp/5de0d21f-605f-403d-bc7e-6b0955f0b78b'...
ssh: Could not resolve hostname github.com: nodename nor servname provided, or not known
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Traceback (most recent call last):
  File "/Users/thead/anaconda/envs/bhub/bin/jupyter-repo2docker", line 11, in <module>
    load_entry_point('jupyter-repo2docker', 'console_scripts', 'jupyter-repo2docker')()
  File "/Users/thead/git/jupyterhub/repo2docker/repo2docker/__main__.py", line 6, in main
    f.start()
  File "/Users/thead/git/jupyterhub/repo2docker/repo2docker/app.py", line 366, in start
    checkout_path
  File "/Users/thead/git/jupyterhub/repo2docker/repo2docker/app.py", line 146, in fetch
    for line in execute_cmd(command, capture=self.json_logs):
  File "/Users/thead/git/jupyterhub/repo2docker/repo2docker/utils.py", line 18, in execute_cmd
    raise subprocess.CalledProcessError(ret, cmd)
subprocess.CalledProcessError: Command '['git', 'clone', '--depth', '50', 'git@github.com:binder-examples/requirements.git', '/tmp/4f9641fa-516c-4d5d-90e8-e56d9f1b7cd1']' returned non-zero exit status 128.
```
So I propose to catch the `CalledProcessError` and raise `RuntimeError('Cloning repository (%s) failed.' % repo)` which we can either catch and nicely format or just show as traceback in the main app.

Reply:
Fair enough! Do you still wanna keep the split up functions though?